openslides/core/static/js/core/site.js

```diff
@@ -952,76 +973,59 @@ angular.module('OpenSlidesApp.core.site', [
+    '$filter',
     '$stateParams',
-    '$location',
-    '$sanitize',
+    'Search',
     'DS',
-    function ($scope, $http, $stateParams, $location, $sanitize, DS) {
-        $scope.fullword = false;
-        $scope.filterAgenda = true;
-        $scope.filterMotion = true;
-        $scope.filterAssignment = true;
-        $scope.filterUser = true;
-        $scope.filterMedia = true;
+    'Motion',
+    function ($scope, $filter, $stateParams, Search, DS, Motion) {
+        $scope.searchresults = [];
+        var searchModules = Search.getAll();
 
         // search function
         $scope.search = function() {
-            var query = _.escape($scope.query);
-            if (query !== '') {
-                var lastquery = query;
-                // attach asterisks if search is not for full words only
-                if (!$scope.fullword) {
-                    if (query.charAt(0) != '*'){
-                        query = "*" + query;
-                    }
-                    if (query.charAt(query.length - 1) != '*'){
-                        query = query + "*";
+            $scope.results = [];
+            var foundObjects = [];
+            // search in rest properties of all defined searchModule
+            // (does not found any related objects, e.g. speakers of items)
+            _.forEach(searchModules, function(searchModule) {
+                var result = {};
+                result.verboseName = searchModule.verboseName;
+                result.collectionName = searchModule.collectionName;
+                result.urlDetailState = searchModule.urlDetailState;
+                result.weight = searchModule.weight;
+                result.checked = true;
+                result.elements = $filter('filter')(DS.getAll(searchModule.collectionName), $scope.searchquery);
+                $scope.results.push(result);
+                _.forEach(result.elements, function(element) {
+                    foundObjects.push(element);
+                });
+            });
+            // search additionally in specific releations of all defined searchModules
+            _.forEach(searchModules, function(searchModule) {
+                _.forEach(DS.getAll(searchModule.collectionName), function(object) {
+                    if (_.isFunction(object.hasSearchResult)) {
+                        if (object.hasSearchResult(foundObjects, $scope.searchquery)) {
+                            // releation found, check if object is not yet in search results
+                            _.forEach($scope.results, function(result) {
+                                if ((object.getResourceName() === result.collectionName) &&
+                                        _.findIndex(result.elements, {'id': object.id}) === -1) {
+                                    result.elements.push(object);
+                                }
+                            });
+                        }
+                    } else {
+                        return false;
```

Comment:
Tested without else-return. Result: It is required. - Otherwise you have to iterate over all users although no users (the users factory) provide the `hasSearchResult`. So the return cancel the iteration already on first run and `if (_.isFunction(object.hasSearchResult))` is false.